code/game/objects/items/weapons/grenades/atmosgrenade.dm

```diff
@@ -0,0 +1,99 @@
+
+
+
+/obj/item/weapon/grenade/gas
+	name = "Plasma Fire Grenade"
+	desc = "A compressed plasma grenade, used to start horrific plasma fires."
+	origin_tech = "materials=3;magnets=4;syndicate=4"
+	icon = 'icons/obj/grenade.dmi'
+	icon_state = "syndicate"
+	item_state = "flashbang"
+	var/gas_type = "fire"
+	var/max_spread_turfs = 20
+	var/gas_debug = 0
+
+/obj/item/weapon/grenade/gas/overpressure
+	name = "Overpressure Grenade"
+	desc = "A compressed gas grenade, used to create crushing pressure in an area."
+	gas_type = "overpressure"
+
+/obj/item/weapon/grenade/gas/suffocation
+	name = "Suffocation Grenade"
+	desc = "An oxygen depletion grenade, used to ensure anyone entering the target area quickly suffocates."
+	gas_type = "suffocation"
+
+/obj/item/weapon/grenade/gas/vacuum
+	name = "Vacuum Grenade"
+	desc = "A grenade that completely removes all air and heat from its detonation area."
+	gas_type = "vacuum"
+
+/obj/item/weapon/grenade/gas/knockout
+	name = "Knockout Grenade"
+	desc = "A grenade that completely removes all air and heat from its detonation area."
+	gas_type = "knockout"
+
+/obj/item/weapon/grenade/gas/prime()
+	var/turf/simulated/origin = get_turf(src)
+	var/list/spread_turfs = list()
+	var/list/scan_turfs = list()
+
+	if(!istype(origin))
+		// Error, unsimulated turf.
+		qdel(src)
+		return
+
+	scan_turfs += origin
+
+	while(spread_turfs.len < max_spread_turfs && scan_turfs.len)
+		if(gas_debug)
+			visible_message("[src] loops: [spread_turfs.len] turfs, [scan_turfs.len] to scan!")
+		var/turf/scan_from = pick(scan_turfs)
+		var/list/adj = 	scan_from.GetAtmosAdjacentTurfs(1)
+		for(var/turf/simulated/C in adj)
+			if(!(C in scan_turfs) && C.CanAtmosPass(C))
+				scan_turfs += C
+		spread_turfs += scan_from
+		scan_turfs -= scan_from
+
+	if(!spread_turfs.len)
+		visible_message("<span class='notice'>[src] detonates harmlessly in the airless space.</span>")
+		qdel(src)
+	else
+		if(gas_debug)
+			visible_message("[src] affects [spread_turfs.len] turfs, with [scan_turfs.len] left to scan!")
+		for(var/turf/simulated/target_turf in spread_turfs)
+			switch(gas_type)
```

Comment:
SNOWFLEKED CODE!!!
===

Instead, create a proc which is overridden by the different grenade subtypes.